core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -211,7 +212,7 @@ public void sendMessage(ReqT message) {
     // For unary requests, we don't flush since we know that halfClose should be coming soon. This
     // allows us to piggy-back the END_STREAM=true on the last message frame without opening the
     // possibility of broken applications forgetting to call halfClose without noticing.
-    if (!unaryRequest) {
+    if (!unaryRequest && !inOnReady) {
```

Comment:
@nmittler, FYI, there are no manual calls to `stream.flush()`. We don't expose flush in our API at the Call level.